extensions/applicationinsights-osplugin-js/src/OsPlugin.ts

```diff
@@ -0,0 +1,257 @@
+                        _mergeOsNameVersion = _ocConfig.mergeOsNameVersion;
+                    } else if (core.getPlugin("Sender").plugin){
+                        _mergeOsNameVersion = true;
+                    } else {
+                        _mergeOsNameVersion = false;
+                    }
+                  
+                    let excludePageUnloadEvents = coreConfig.disablePageUnloadEvents || [];
+
+                    if (_excludePageUnloadEvents && _excludePageUnloadEvents !== excludePageUnloadEvents) {
+                        removePageUnloadEventListener(null, _evtNamespace);
+                        removePageHideEventListener(null, _evtNamespace);
+                        _excludePageUnloadEvents = null;
+                    }
+                   
+                    if (!_excludePageUnloadEvents) {
+                        // If page is closed release queue
+                        addPageUnloadEventListener(_doUnload, excludePageUnloadEvents, _evtNamespace);
+                        addPageHideEventListener(_doUnload, excludePageUnloadEvents, _evtNamespace);
+                    }
+                    _excludePageUnloadEvents = excludePageUnloadEvents;
+                }));
+                function _doUnload() {
+                    _releaseEventQueue();
+                }
+            };
+        
+            _self.processTelemetry = (event: ITelemetryItem, itemCtx?: IProcessTelemetryContext) => {
+                itemCtx = _self._getTelCtx(itemCtx);
+
+                if (!_retrieveFullVersion && !_getOSInProgress && _firstAttempt) {
+                    // Start Requesting OS version process
+                    _getOSInProgress = true;
+                    startRetrieveOsVersion();
+                    _firstAttempt = false;
+                }
+        
+                if (_getOSInProgress) {
+                    _eventQueue.push({
+                        ctx: itemCtx,
+                        item: event
+                    });
+                } else {
+                    updateTeleItemWithOs(event);
+                    _self.processNext(event, itemCtx);
+                }
+            };
+
+            _self._doTeardown = (unloadCtx?: IProcessTelemetryUnloadContext, unloadState?: ITelemetryUnloadState) => {
+                _completeOsRetrieve();
+                removePageUnloadEventListener(null, _evtNamespace);
+                removePageHideEventListener(null, _evtNamespace);
+                // Just register to remove all events associated with this namespace
+                _initDefaults();
+            };
+
+        
+            /**
+             * Wait for the response from the browser for the OS version and store info in the session storage
+             */
+            function startRetrieveOsVersion() {
+                // Timeout request if it takes more than 5 seconds (by default)
+                _getOSTimeout = setTimeout(() => {
```

Comment:
For this we can now use the `scheduleTimeout` which returns an `ITimerHandler`

Reply:
It then has a `cancel` function (instead of calling `clearTimeout` and you can just reset to null when done.